Core review breakpoint

SirMorek opened the pull request:
Almost complete, I think I'm just missing packaging and deployment.

SirMorek committed:
Initial commit

SirMorek committed:
Matchmaker: Add model for hockey players

SirMorek committed:
Matchmaker: Add first assignment module

This is a 'dummy' assigner who minimally assigns teams but does not attempt
to create balanced teams beyond headcount.

SirMorek committed:
Matchmaker: First pass complete of matchmaker

Matchmaker is now a walking skeleton. Arbitrary lists of players and requested team counts
can be provided and matches will be provided such that teams are equal in headcount and
the remainder returned as well.

Full completion of the acceptance criteria will require implementing alternative assignment
algorithms, but this is enough that we can start working on other parts of the solution.

SirMorek committed:
Matchmaker: Add repr's for players and teams

These are convenient for development, as we work towards on the presentation layer.

SirMorek committed:
Webservice: Initial commit

SirMorek committed:
Matchmaker: Add serialization to teams and hockey players

This will help for persisting information in the webserver, going for simple
file storage.

SirMorek committed:
Webservice: Add file level caching for basic persistence

This gets the job done with carrying forward the generated teams
until they are requested to be regenerated. That will be as simple
as nuking the file and re-running the matchmaker with new params.

SirMorek committed:
Webservice templates: Add tabular presentation of players

SirMorek committed:
Matchmaker/Web-service: Add team skill averages

SirMorek committed:
Webservice: Refactor app to add generate/reset buttons

SirMorek committed:
Webservice: Add form for controlling number of teams

SirMorek committed:
Matchmaker: Implement a greedy assigner

SirMorek committed:
Matchmaker: Implement a tree search assigner

SirMorek committed:
Squadmaker: Add debug script to compare performance of different assigners

SirMorek committed:
Matchmaker: Switch main algorithm over to tree search

gralamin commented on matchmaker/player.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def get_skills(player_json):
+    return {
```
Curly braces aren't required, Round brackets `()` would be fine.

SirMorek replied:
I wanted a dict comprehension, i believe that you are talking about something else...

gralamin commented on matchmaker/player.py:
```diff
@@ -0,0 +1,49 @@
+'''
+Abstract Player.
```
Docstring saids Abstract here, but the Player class isn't using abc. Just say Base class?

gralamin commented on matchmaker/player.py:
```diff
@@ -0,0 +1,49 @@
+
+
+class HockeyPlayer(Player):
+    SKILLS = ['shooting', 'skating', 'checking']
```
If you are using python 3, consider using an enum.

SirMorek replied:
Should just delete it, its not used or useful. An enum is probably better for what it was supposed to do, you are correct.

gralamin commented on matchmaker/assign.py:
```diff
@@ -0,0 +1,34 @@
+
+class RoundRobinAssigner(object):
+    '''
+    A basic team builder that simply round-robins through the provided players
```
This doesn't look like its pep257 compliant.

SirMorek replied:
Could you be more specific about what is in violation here? Reading through pep257 the main thing it would be unhappy with is my use of ''' instead of """

gralamin commented on matchmaker/assign.py:
```diff
@@ -0,0 +1,34 @@
+'''
+
+
+class RoundRobinAssigner(object):
```
There currently isn't any reason for this to be a class, maybe future commits will make a reason, but unless you want to enforce the same interface on multiple things (which I would then expect an abstract base class), its not a great sign.

SirMorek replied:
Fair, just yielding the methods would have the same outcome. You see here the same problem I had with player where I started the project thinking I wanted ABC patterns but didn't follow through.

gralamin commented on matchmaker/team.py:
```diff
@@ -0,0 +1,19 @@
+'''
+
+
+class Team(object):
```
You added Teams in this commit, but didn't add any tests for them.

gralamin commented on matchmaker/main.py:
```diff
@@ -0,0 +1,16 @@
+'''
+The matchmaker service is responsible for taking the set of players and
```
pep257 note again.

gralamin commented on matchmaker/unittest/test_main.py:
```diff
@@ -0,0 +1,10 @@
+
+def test_matchmake():
+    import json
+    with open("../players.json") as player_file:
```
Generally you don't want to open non-test files in unit tests.

SirMorek replied:
Yeah I should replace that dependency with a new one generated just for testing...

gralamin commented on web-service/matchmaker:
```diff
@@ -0,0 +1 @@
+../matchmaker/
```
Such a dirty hack. Hoping this gets fixed later.

gralamin commented on web-service/app.py:
```diff
@@ -10,6 +12,36 @@
+        teams, players = matchmake(player_json['players'], 6)
 
-    return "Teams: %s\nRemainder: %s" % (teams, remainder)
+    with open("persistence.json", "w") as outfile:
```
Consider separating some of this logic into its own function to enable easy testing without worrying about the Flask part.

gralamin commented on web-service/static/style.css:
```diff
@@ -0,0 +1,134 @@
+html {
```
This is a sort of common thing - treating the UI related changes as not worth breaking apart. I suggest talking in the commit message a bit about why you choose some of the styling, even if it works out to something like "stole bootstrap.css".

gralamin commented on web-service/templates/base.html:
```diff
@@ -1,20 +1,58 @@
   </head>
   <body>
+    <style>
+      table, th, td {
```
You have a css document, why put this stuff here instead of applying it to a class or something there?

SirMorek replied:
Legit tried and it refused to work. I don't understand why doing it here worked and doing the same stuff in the stylesheet didn't, so I cut my losses with it and moved on to other parts of the problem.

gralamin commented on web-service/app.py:
```diff
@@ -1,33 +1,19 @@
 import json
-from flask import Flask, render_template
+from os import remove
```
This commit ("Webservice: Refactor app to add generate/reset buttons" is doing a bunch of unrelated refactoring work (making a PERSISTENCE_FILE variable for example). Consider splitting up.

gralamin commented on matchmaker/assign.py:
```diff
@@ -4,6 +4,17 @@
+from copy import deepcopy
+
+
+def evaluate_SSE(teams):
```
Isn't clear what SSE means here.

SirMorek replied:
Agreed, worth including a docstring. (Sum of squared errors)

gralamin reviewed (commented):
You mentioned you didn't spend time for packaging and deployment ahead of time. I think you would have benefited quite a bit for setting up your environment to deploy any time, with scripts for flake8, flake8-docstrings, etc.

That said, my first round of comments aren't too bad, and some of them are probably ignorable. The design as a whole looks alright - it could be improved, but it seems like it would work. The algorithms choosen appear to be tested in such a way that you understand the consequences of your algorithm,.

SirMorek committed:
Matchmaker: Initial python project structure

SirMorek committed:
Matchmaker: Add a sample player file for unit testing

SirMorek committed:
Webservice: Add stylesheet

Copied from the flask tutorial at https://github.com/pallets/flask/tree/master/examples/tutorial/

SirMorek committed:
Webservice: Add initial skeleton for templated html generation

SirMorek commented:
Making a new PR to show the latest tip.
